update-index.nix

```diff
@@ -0,0 +1,28 @@
+{
+  pkgs ? import <nixpkgs> {},
+
+  coreutils ? pkgs.coreutils,
+  gnugrep ? pkgs.gnugrep,
+  lib ? pkgs.lib,
+  nix-index ? pkgs.nix-index,
+  writeScript ? pkgs.writeScript,
+}:
+
+writeScript "update-index" ''
+  PATH=${lib.makeBinPath [
+    coreutils
+    gnugrep
+    nix-index
+  ]}
+
+  # on flake based installations nixpkgs is specified via
+  # flake input and therefore NIX_PATH might be unset
+  if echo $NIX_PATH | grep -q "nixpkgs="; then
+    nixpkgs=""
+  else
+    nixpkgs="-I nixpkgs=${pkgs.path}"
```

Comment:
This line causes the entire 1+GB nixpkgs checkout to be copied to remote builders on every build which is really really annoying.

Reply:
I guess the whole remote build protocol itself is not optimal. It doesn't do any kind of compression when sending data (AFAIK). If only openssh would support any kind of useful compression (gzip is not), then the whole thing would be a lot more usable. But sadly they don't get that done despite highly efficient real time compression algos like lz4 exist already since more than 10 years.

OK, but enough blaming others. What can we do to improve this?
We could allow the nixpkgs being `null` at build time in which case the `update-index` script would fall back to `NIX_PATH` at runtime. You then just would have to make sure to set `NIX_PATH` accordingly.

Or we remove the `update-index` completely, forcing the user to use the nix-index tool manually. But then there is no guarantee that the index is generated against any kind of recent nixpkgs version.

Reply:
In your specific case you could probably do some `builtins.path` + `filter` magic to get rid of the `.git` directory. Also enabling `gzip` in you ssh config might increase the speed, but since its CPU bound nature it will also limit your bandwidth to a few MB/s.

Reply:
> I guess the whole remote build protocol itself is not optimal. It doesn't do any kind of compression when sending data (AFAIK). If only openssh would support any kind of useful compression (gzip is not), then the whole thing would be a lot more usable. But sadly they don't get that done despite highly efficient real time compression algos like lz4 exist already since more than 10 years.

I don't think there is much that openssh can do with encryption and nix streaming data over ssh. Nix should just locally compress things before copying it to the remote worker.

> In your specific case you could probably do some builtins.path + filter magic to get rid of the .git directory.

It turns out if I use comma via the flake nix does that already and the size is reduced to 10% which is still over 100MB+.

I ended up doing

```nix
comma = import comma { pkgs = pkgs // { path = "/dev/null"; };
```

which also removed the last 100MB+.